### PR TITLE
Get issued tokens info

### LINKED ISSUE #3: Token info
Add api route that takes array of token ids and returns boxes, where this tokens were issued.

### PR DESCRIPTION
**Pull request by greenhat**
Close #3 

Repo:
- [x] implement `AssetRepo.getIssuingBoxes(tokenIds)`; 
 
Add REST API calls:
- [x] implement `AssetsService.getAllIssuingBoxes`;
- [x] implement `AssetsService.getIssuingBoxes()`;
- [x] add routes;

### CODE REVIEW

File: modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/repositories/AssetRepo.scala [hunk 1]
```diff
@@ -36,6 +37,16 @@ trait AssetRepo[D[_], S[_[_], _]] {
     offset: Int,
     limit: Int
   ): S[D, Address]
+
+  /** Get boxes where all tokens where issued
+    * according to EIP-4 https://github.com/ergoplatform/eips/blob/master/eip-0004.md
+    */
+  def getAllIssuingBoxes: D[List[ExtendedOutput]]
```
Comment (oskin1) [inline modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/repositories/AssetRepo.scala]: I think it's better to wrap the result into streaming effect here, as the result set can potentially not fit into list.
`S[D, ExtendedOutput]`

File: modules/explorer-core/src/test/scala/org/ergoplatform/explorer/db/repositories/AssetRepoSpec.scala [hunk 2]
```diff
@@ -33,8 +34,33 @@ class AssetRepoSpec
...
+
   private def withLiveRepo[D[_]: LiftConnectionIO: Sync](
     body: AssetRepo[D, fs2.Stream] => Any
   ): Any =
     body(db.repositories.AssetRepo[D])
+
+  private def withLiveRepos[D[_]: LiftConnectionIO: Sync](body:
+                                                          (AssetRepo[D, fs2.Stream],
+                                                            OutputRepo[D, fs2.Stream],
+                                                            InputRepo[D]) => Any): Any =
```
Comment (oskin1) [inline modules/explorer-core/src/test/scala/org/ergoplatform/explorer/db/repositories/AssetRepoSpec.scala]: Please use scala-fmt to format code according to project configuration

Comment (greenhat) [reply modules/explorer-core/src/test/scala/org/ergoplatform/explorer/db/repositories/AssetRepoSpec.scala]: Thanks, fixed. Although I enabled scalafmt in IDEA, it still uses its own formatter for line range by default. 🤦‍♂

File: modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/repositories/AssetRepo.scala (hunk 1 above)
Comment (greenhat) [reply modules/explorer-core/src/main/scala/org/ergoplatform/explorer/db/repositories/AssetRepo.scala]: Done.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/routes/AssetsRoutes.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+) {
+
+  import org.ergoplatform.explorer.http.api.v0.defs.AssetsEndpointDefs._
+
+  val routes: HttpRoutes[F] =
+    getAllIssuingBoxesR
+
+  private def getAllIssuingBoxesR: HttpRoutes[F] =
+    getAllIssuingBoxesDef.toRoutes { _ =>
+      service.getAllIssuingBoxes.compile.toList.either
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/routes/AssetsRoutes.scala]: Why compiling stream here, we are loosing all gained earlier benefits doing this. Http4s supports streaming.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      .create[F]
+      .map { implicit logger =>
+        new Live(
+          AssetRepo[D]
+        )(xa)
+      }
+
+  final private class Live[
+    F[_]: Sync: Logger,
+    D[_]: ContravariantRaise[*[_], InconsistentDbData]: Monad
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala]: Seems you don't raise `InconsistentDbData` anywhere in the service, why do you need this type-class then?

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    Slf4jLogger
+      .create[F]
+      .map { implicit logger =>
+        new Live(
+          AssetRepo[D]
+        )(xa)
+      }
+
+  final private class Live[
+    F[_]: Sync: Logger,
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala]: You don't use neither `Sync` nor `Logger` functionality on `F[_]`.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala (hunk 4 above)
Comment (greenhat) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala]: It was a copy-paste leftover. Now, I do use it in `getIssuingBoxes`.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala (hunk 5 above)
Comment (greenhat) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/services/AssetsService.scala]: Thanks! Removed.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/routes/AssetsRoutes.scala (hunk 3 above)
Comment (greenhat) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/routes/AssetsRoutes.scala]: I saw [`// API v0 format does not allow to use streaming`](http://github.com/ergoplatform/explorer-backend/blob/bb967e951790aa44da188b962953da034980f8df/modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/routes/BlocksRoutes.scala#L30-L30) comment. There are no endpoints with streaming in the code and I could not make one.

**Comment by greenhat** [comment]
@oskin1 I'm finished with this PR. One caveat though, I did not implement API call for multiple token ids, only for one token id. Cause I did not found anywhere in our API (old explorer) how we want to deal with an array of values as a query parameter. I suggest leave it for later.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/routes/AssetsRoutes.scala (hunk 3 above)
Comment (oskin1) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v0/routes/AssetsRoutes.scala]: Right, many of existing methods from APIv0 doesn't allow to use streaming due to response structure, but you're adding new method to APIv0 and streaming can easily be used here.
You can find answers here https://tapir-scala.readthedocs.io/en/latest/server/http4s.html

**Pull request merged**